sdk/core/allocator/main.cc

```diff
@@ -732,12 +735,22 @@ namespace
 			{
 				return 0;
 			}
-			owner.quota += chunk.size_get();
+			size_t chunkSize = chunk.size_get();
+			chunk.ownerID    = 0;
 			if (chunk.claims == 0)
 			{
-				return gm->mspace_free(chunk, bodySize);
+				int ret = gm->mspace_free(chunk, bodySize);
+				// If free fails, don't manipulate the quota.
+				if (ret == 0)
+				{
+					owner.quota += chunkSize;
+				}
+				return ret;
 			}
-			chunk.ownerID = 0;
+			// We've removed the owner, so refund the quota immediately, the
+			// free won't happen until the last claim goes away, but this is no
+			// longer the owner's responsibility.
```

Comment:
A bit unfortunate that this means there is memory that is not free (could remain allocated indefinitely) but isn't attributed to a quota. With the existing claim mechanism does it get charged to the claimer's quota?